examples/timeModels.cc

```diff
@@ -52,9 +52,9 @@ double buildModelsImpl(
     ndarray::Array<T,2,-2> matrix = matrixT.transpose();
     double result = 0.0;
     for (ndarray::Array<double const,2>::Iterator i = parameters.begin(); i != parameters.end(); ++i) {
-        ellipse.setParameterVector(i->asEigen());
+        ellipse.setParameterVector(ndarray::asEigenMatrix(*i));
```

Comment:
You could avoid the awkward `*i` with a foreach loop.

Reply:
I would hold off on this for now, or at least be very careful testing that change.  This just reminded me that https://github.com/ndarray/ndarray/blob/master/include/ndarray/detail/NestedIterator.h#L43 might need to be changed to the InputIterator concept equivalent, and not doing so might get in the way of higher-dimensionsal ndarray iterators working properly in range-based for.

Long version of the story is that proxy iterators in C++ aren't allowed to be RandomAccessIterators (or even ForwardIterators), due to a really annoying edge case in the standard.  Once upon a time, Boost.Iterator and it's not-quite-the-same-as-the-standard's iterator categories were seen as the way towards a future version of the standard without those (and other problems) and that's why I used them here.  AFAIK, that idea has gone by the wayside - I haven't seen anything about fixing proxy iterators in the latest standards (though I haven't paid super close attention) - and I think we just need to stop promising that these are RandomAccessIterators, even though they can do 99% of the things a RandomAccessIterator can do.

Sorry to dump that on you right before I disappear for a week, and I'm even more sorry I didn't fix this earlier - it *is* a serious problem, but I keep remembering it right when I can't fix it immediately.  But for now, stay away from range-based `for` on 2-d or higher ndarray iterators.

Reply:
Sorry, I don't understand: as far as I can tell for-each loops should work just fine on input iterators, since they only need inequality, increment, and dereference capability. They certainly don't need random access.

Reply:
I'll revert the change.

Reply:
Yes, for-each will work on either kind of iterator, but the compiler is allowed to do different things under the hood in the two cases, and the code it generates for random-access iterators may yield incorrect behavior if the iterator really only models input iterator.  I have never really nailed down what is going on here (doing that is part of why I never got around to fixing it, though I really should just do it and investigate later), but I have seen at least one case (in an early version of `afw::geom::SpanSet`) where having an incorrect category for a proxy iterator led to some weird behavior in a range-based `for`, and that weird behavior went away if we either fixed the category or switched to a regular `for` loop (obviously the former is the right fix, but it always bothered me that I couldn't figure out what the range-based `for` was doing to cause trouble).